bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+                using MP = macroni::macroni::MacroParameter;
+                for_op.getCondRegion().walk(
+                    [&](Op *op2) {
+                        if (auto param_op = mlir::dyn_cast<MP>(op2)) {
+                            if (param_op.getParameterName() == "pos") {
+                                list_for_each_to_pos[op] = op2;
+                            } else if (param_op.getParameterName() == "head") {
+                                list_for_each_to_head[op] = op2;
+                            }
+                        }
+                    }
+                );
+                return !(list_for_each_to_pos.contains(op) &&
+                         list_for_each_to_head.contains(op));
+            });
+        trg.addDynamicallyLegalOp<vast::hl::CallOp>(
+            [](Op *op) {
+                using CO = vast::hl::CallOp;
+                CO call_op = mlir::dyn_cast<CO>(op);
+                llvm::StringRef name = call_op.getCalleeAttr().getValue();
+                if (rcu_read_unlock != name) {
+                    return true;
+                }
+                llvm::APInt unlock_level = get_lock_level(op);
+                for (auto cur = op; cur; cur = cur->getPrevNode()) {
+                    if (CO call_op = mlir::dyn_cast<CO>(cur)) {
+                        name = call_op.getCalleeAttr().getValue();
+                        if (rcu_read_lock == name) {
+                            llvm::APInt lock_level = get_lock_level(cur);
+                            if (unlock_level == lock_level) {
+                                unlock_to_lock[op] = cur;
+                                break;
+                            }
+                        }
+                    }
+                }
+                return !unlock_to_lock.contains(op);
+            });
+        trg.markUnknownOpDynamicallyLegal([](Op *op) { return true;});
+
+        mlir::RewritePatternSet patterns(&*mctx);
+        patterns.add<macroni::macro_expansion_to_get_user>(patterns.getContext());
+        patterns.add<macroni::macro_expansion_to_offsetof>(patterns.getContext());
+        patterns.add<macroni::macro_expansion_to_container_of>(patterns.getContext());
+        patterns.add<macroni::macro_expansion_to_rcu_dereference>(patterns.getContext());
+        patterns.add<macroni::macro_expansion_to_smp_mb>(patterns.getContext());
+        patterns.add<macroni::for_to_list_for_each>(patterns.getContext());
+        patterns.add<macroni::rcu_read_unlock_to_rcu_critical_section>(patterns.getContext());
+        Op *mod_op = mod.get().getOperation();
+        // Apply the conversions. Cast the result to void to ignore no_discard
+        // errors
+        (void) mlir::applyPartialConversion(mod_op, trg, std::move(patterns));
+
+        // Print the result
+        mlir::OpPrintingFlags flags;
+        flags.enableDebugInfo(false, true);
+        mod->print(llvm::outs(), flags);
+
+        return EXIT_SUCCESS;
```

Comment:
I may be reading it incorrectly, but currently it seems you are exiting after the first job is processed? Maybe add a comment why or drop the for-loop.